src/org/zaproxy/zap/extension/frontEndScanner/ExtensionFrontEndScanner.java

```diff
@@ -0,0 +1,328 @@
+        return Constant.ZAP_TEAM;
+    }
+
+    @Override
+    public String getDescription() {
+        return Constant.messages.getString(PREFIX + ".desc");
+    }
+
+    @Override
+    public URL getURL() {
+        try {
+            return new URL(Constant.ZAP_EXTENSIONS_PAGE);
+        } catch (MalformedURLException e) {
+            return null;
+        }
+    }
+
+    @Override
+    public boolean onHttpRequestSend(HttpMessage msg) {
+        return true;
+    }
+
+    @Override
+    public boolean onHttpResponseReceive(HttpMessage msg) {
+        String host = msg.getRequestHeader().getHeader("host");
+        String frontEndApiUrl = API
+            .getInstance()
+            .getCallBackUrl(this.api, new String("https://" + host));
+
+        if (frontEndScannerEnabled && msg.getResponseHeader().isHtml()) {
+            try {
+                String html = msg.getResponseBody().toString();
+
+                Source document = new Source(html);
+                List<Element> heads = document.getAllElements("head");
+                Element head = heads.isEmpty() ? null : heads.get(0);
+
+                if (head != null) {
+                    int historyReferenceId = msg.getHistoryRef().getHistoryId();
+
+                    String injectedContent =
+                        "<script type='text/javascript'>"
+                        + "var frontEndScanner=(function() {"
+                        +     "const HISTORY_REFERENCE_ID = " + historyReferenceId + ";"
+                        +     "const CALLBACK_ENDPOINT = '" + frontEndApiUrl + "';"
+                        +     userScriptsToInject()
+                        +     frontEndScannerCode()
+                        + "})();"
+                        + "</script>";
+
+                    OutputDocument newResponseBody = new OutputDocument(document);
+                    int insertPosition = head.getChildElements().get(0).getBegin();
+                    newResponseBody.insert(insertPosition, injectedContent);
+
+                    msg.getResponseBody()
+                        .setBody(newResponseBody.toString());
+
+                    int newLength = msg.getResponseBody().length();
+                    msg.getResponseHeader().setContentLength(newLength);
+                } else {
+                    LOGGER.error("<head></head> is missing in the response");
+                }
+            } catch (Exception e) {
+                LOGGER.error(e.getMessage(), e);
+            }
+        }
+        return true;
+    }
+
+    @Override
+    public int getArrangeableListenerOrder() {
+        // Need to run after the HistoryReference has been saved to the database
+        return ProxyListenerLog.PROXY_LISTENER_ORDER + 42;
+    }
+
+    private String frontEndScannerCode() throws IOException {
+        Path frontEndScannerPath = Paths.get(FRONT_END_SCANNER);
+        return readFromFile(frontEndScannerPath);
+    }
+
+    private String userScriptsToInject() throws IOException {

```

Comment:
Should read these in from the scripts extension. That way people can enable and disable the scripts individually: https://github.com/zaproxy/zaproxy/blob/develop/src/org/zaproxy/zap/extension/pscan/scanner/ScriptsPassiveScanner.java#L79-L101

Reply:
Done!